### PR TITLE
Add JNI activation and library path management

### LINKED ISSUE #2999: Offer a JNIBuildItem to let extensions enable JNI
See https://groups.google.com/d/topic/quarkus-dev/pLwbEB1VCP0/discussion for the conversation

### PR DESCRIPTION
**Pull request by Dufgui**
fix #2999

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java]: Build items need to be immutable. The producer/consumer model does not really allow for mutation, so this might work if the steps that modify it also produce other items and the order happens to work out, however it is a bug waiting to happen.

Instead this should be a MultiBuildItem and extensions should just produce an instance if they require JNI. If this list is empty then nothing requires JNI, otherwise it should be enabled.

File: extensions/jni/deployment/src/main/java/io/quarkus/jni/deployment/JniProcessor.java
Comment (stuartwdouglas) [inline extensions/jni/deployment/src/main/java/io/quarkus/jni/deployment/JniProcessor.java]: This should not have its own extension. If a user wants to enable JNI they should be able to do it through application.properties. If an extension needs it they should be able to just produce a build item.

**Review by stuartwdouglas — DISMISSED** [review]

**Comment by Dufgui** [comment]
I see this commit:
https://github.com/oracle/graal/commit/cbadcf6f7a1890562984e1c581f7c3b31809b34e#diff-c643013b1a9b732d05431357cc5ec432

So now I uncomment the code which push many lib path.
Now I will test this

**Comment by Dufgui** [comment]
So I add manually 
`args.add("-Djava.library.path=.");`
but I continue to have the following error. Even if the native image compilation is done with jni flag.
Any idea of methodology to find my mistake ? Because, i don't see any error...


Caused by: java.lang.UnsatisfiedLinkError: org.acme.JniWrapper.getString()Ljava/lang/String; [symbol: Java_org_acme_JniWrapper_getString or Java_org_acme_JniWrapper_getString__]
	at com.oracle.svm.jni.access.JNINativeLinkage.getOrFindEntryPoint(JNINativeLinkage.java:145)

File: `
Comment (machi1990) [inline `]: This should not be committed.

**Comment by Dufgui** [comment]
Test seems ok, sorry it's my fault, my test are wrong previously. My system load are in static block. I win  🥇 of the noob ;o)

**Comment by Dufgui** [comment]
https://github.com/quarkusio/quarkus-quickstarts/pull/235
here is my test, works on mac, not test on other OS

File: extensions/jni/deployment/src/main/java/io/quarkus/jni/deployment/JniProcessor.java
Comment (Dufgui) [reply extensions/jni/deployment/src/main/java/io/quarkus/jni/deployment/JniProcessor.java]: Ok so I move Processor to the core

Comment (Dufgui) [reply extensions/jni/deployment/src/main/java/io/quarkus/jni/deployment/JniProcessor.java]: done

File: core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java]: Wondering if it should throw an exception instead?

If we decide to throw an exception, we could simplify this code a bit and produce the build item outside of the if.

**Review by gsmet — COMMENTED** [review]
I think we're on the right path, thanks for this.

Could you take a look at my comments?

File: core/deployment/src/main/java/io/quarkus/deployment/JniProcessor.java
Comment (machi1990) [inline core/deployment/src/main/java/io/quarkus/deployment/JniProcessor.java]: I think there is no need for `name` since it is the same as field name. It will be auto picked from the field name.

Comment (Dufgui) [reply core/deployment/src/main/java/io/quarkus/deployment/JniProcessor.java]: You are right ;o)
by default it's `ConfigItem.HYPHENATED_ELEMENT_NAME` which is a the field name -> camel case + hyphnated. Thanks now I discover the process `ConfigDefinition.processConfigGroup`

File: core/creator/src/main/java/io/quarkus/creator/phase/nativeimage/NativeImagePhase.java
Comment (machi1990) [inline core/creator/src/main/java/io/quarkus/creator/phase/nativeimage/NativeImagePhase.java]: Extract a variable for `properties.getProperty("quarkus.jni.enable")`maybe?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java]: I don't think we should merge the two ifs. The logic is totally different.

Can you move the JNI stuff to a separate if?

**Review by gsmet — CHANGES_REQUESTED** [review]
@Dufgui I think it looks good now. I added a comment about the organization of part of the code. Could you take a look?

**Comment by gsmet** [comment]
@Dufgui and once you're done, can you squash all your commits (and fix the comment as we don't have an extension anymore)

Thanks!

File: core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java]: Thinking more about it, I have a question about this: do we really need to make the path absolute?

Because it will make it quite difficult to build on one machine and push to another?

**Comment by Dufgui** [comment]
I do it because jdk do it too in runtime class:

if (!(new File(filename).isAbsolute())) {
  789               throw new UnsatisfiedLinkError(
  790                   "Expecting an absolute path of the library: "
+ filename);  791           }

If you préfère I can let them pass and have an exception



Le sam. 6 juil. 2019 à 16:46, Guillaume Smet <notifications@github.com> a
écrit :

**Comment by Dufgui** [comment]
Oh sorry you are right. I made a mistake. A mix between system load and
load library


Le sam. 6 juil. 2019 à 16:56, Guillaume Dufour <guillaume.duff@gmail.com> a
écrit :

File: core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/SubstrateConfigBuildStep.java]: Hmmm, do we need `enableJni`? I mean it will always be true if `jniBuildItems` is not empty.

**Review by gsmet — COMMENTED** [review]
@Dufgui sorry, I just noticed one last thing.

Once fixed, I think we are good to go!

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/JniBuildItem.java]: I think we also need an empty constructor as you might want to enable JNI even without any library path (this is the case for the Tika extension for instance).

**Comment by Dufgui** [comment]
done @gsmet

**Comment by gsmet** [comment]
I pushed a couple of very minor changes to your PR. I'm waiting for CI to go green before merging.

**Review by gsmet — APPROVED** [review]
Waiting for CI, thanks!

**Comment by gsmet** [comment]
@gunnarmorling you should be able to use this in your extension now.

**Comment by gunnarmorling** [comment]
Thanks @gsmet, I'll take a look.

**Pull request merged**
